_posts/2020-11-04-solidity-ama-1-recap.md

```diff
@@ -0,0 +1,140 @@
+
+#### How does it work?
+
+There are several steps that the optimizer performs. A simple example would be evaluating expressions whose value is known at runtime, e.g., `x + 0` is evaluated to `x`, where `x` may be a parameter known only during runtime. A more complex example would be to identify expressions that remains invariant inside a loop and moving them outside the loop, thereby saving gas. Another interesting example would be avoiding accessing the same value from storage multiple times, i.e., multiple `sload` to the same slot can be reduced to a single `sload`, in certain situations.
+
+References:
+- [High level description of steps in Yul optimizer](https://github.com/ethereum/solidity/blob/develop/libyul/optimiser/README.md)
+- [List of steps in Yul optimizer](https://solidity.readthedocs.io/en/v0.7.4/yul.html?highlight=ir%20optimize#yul-optimizer)
+- [Assembly based optimizer](https://solidity.readthedocs.io/en/v0.7.4/internals/optimiser.html)
+<br><br>
+
+#### What are the typical differences between optimized and non-optimized code?
+
+Generally, the most visible difference would be constant expressions getting evaluated. When it comes to the ASM output, one can also notice reduction of equivalent/duplicate "code blocks" (compare the output of the flags `--asm` and `--asm --optimize`). However, when it comes to the Yul/intermediate-representation, there can be significant differences, for example, functions can get inlined, combined, rewritten to eliminate redundancies, etc. (compare the output between the flags `--ir` and `--optimize --ir-optimized`).
+
+#### How is optimization affected by the number of runs (`--optimize-runs`) and is there a maximum number above which it stops mattering, or is `--optimize-runs=20000` less efficient than `--optimize-runs=500000`?
+
+The parameter specifies roughly how often each opcode of the deployed code will be executed across the life-time of the contract. A "runs" parameter of "1" will produce long but cheap code. The largest value is ``2**64-1``.
+
+#### When multiple small variables are packed into a single storage slot, how can we be sure accessing them both is done with a single SSTORE/SLOAD? How "nearby" do the accesses have to be each-other, and is there any better way to ensure this than inspecting the assembly?
+
+As long as this still uses the non-yul code generator, it is actually rather limiting. It is best to not have any branches in between the accesses, so assigning a memory struct to storage should work best.
+
+#### Does it make sense to cache ``arr.length`` in a stack variable when looping over an array from front to back, or will it SLOAD/MLOAD that only once if it sees it won't change in the loop body?
+
+In the Yul optimizer, there is an optimization step called **LoopInvariantCodeMotion** designed to detect expressions that remain invariant in the loop and move them outside the loop. Take the following solidity example that finds the sum of a dynamic integer array.
```

Comment:
We should stress that this section deals about the new code generator that is not active yet. For the old generator, there are some inefficiencies and it probably is cheaper to cache the length on the stack.

Reply:
@franzihei The answer was slightly edited after posting to reddit. The text here is hackmd version https://www.reddit.com/r/ethdev/comments/jigz5o/ask_the_solidity_team_anything_1/gahzrhm/?utm_source=reddit&utm_medium=web2x&context=3 Can you please update it?

Reply:
Here is the markdown source: https://gist.github.com/hrkrshnn/a1165fc31cbbf1fae9f271c73830fdda

Reply:
adjusted